pmd-java/src/main/java/net/sourceforge/pmd/lang/java/typeresolution/ClassTypeResolver.java

```diff
@@ -358,18 +386,174 @@ public Object visit(ASTName node, Object data) {
+
+                        previousType = getBestMethodReturnType(methods, astArgumentList, null);
+                    } else { // field
+                        previousType = getFieldType(previousType, dotSplitImage[i], accessingClass);
+                    }
                 }
 
                 if (previousType != null) {
                     node.setTypeDefinition(previousType);
                 }
             }
         }
-
         return super.visit(node, data);
     }
 
+    public JavaTypeDefinition getBestMethodReturnType(List<MethodType> methods, ASTArgumentList arguments,
+                                                      List<JavaTypeDefinition> typeArgs) {
+        // TODO: add overload resolution
+        if (methods.size() == 1) {
+            return methods.get(0).getReturnType();
+        } else {
+            return null;
+        }
+    }
+
+    /**
+     * Search outwards from a node the enclosing type declarations. Searching them and their supertypes
+     * for method
+     */
+    private List<MethodType> getLocalApplicableMethods(TypeNode node, String methodName,
+                                                       List<JavaTypeDefinition> typeArguments,
+                                                       int argArity,
+                                                       Class<?> accessingClass) {
+        List<MethodType> foundMethods = new ArrayList<>();
+
+        if (accessingClass == null) {
+            return foundMethods;
+        }
+
+        // we search each enclosing type declaration, looking at their supertypes as well
+        for (node = getEnclosingTypeDeclaration(node); node != null;
+             node = getEnclosingTypeDeclaration(node.jjtGetParent())) {
+
+            getApplicableMethods(node.getTypeDefinition(), methodName, typeArguments, argArity, accessingClass,
+                                 foundMethods);
+        }
+
+        // TODO: search static methods
+
+        return foundMethods;
+    }
+
+    public List<MethodType> getApplicableMethods(JavaTypeDefinition context,
+                                                 String methodName,
+                                                 List<JavaTypeDefinition> typeArguments,
+                                                 int arity,
+                                                 Class<?> accessingClass) {
+        List<MethodType> result = new ArrayList<>();
+
+        getApplicableMethods(context, methodName, typeArguments, arity, accessingClass, result);
+
+        return result;
+    }
+
+    public void getApplicableMethods(JavaTypeDefinition context,
+                                     String methodName,
+                                     List<JavaTypeDefinition> typeArguments,
+                                     int argArity,
+                                     Class<?> accessingClass,
+                                     List<MethodType> result) {
```

Comment:
not a big fan of this... we have an overloaded version of the method that receives an out parameter instead of returning it as the other one / related methods. This is inconsistent and makes maintenance harder